<a href="https://colab.research.google.com/github/nanomelk/Analista-de-Datos-II/blob/main/ABP2_Home_Credit_Default_Risk.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **EVIDENCIA 1 - Comprensión del Negocio - Datos - Análisis exploratorio**

# **1. COMPRENSIÓN DEL PROBLEMA DE NEGOCIO**

### **Contexto Financiero y Necesidad**
En el sector crediticio, la evaluación precisa del riesgo es fundamental para la sostenibilidad y rentabilidad de la entidad. Otorgar créditos a clientes que posteriormente incurren en mora genera pérdidas financieras directas por capital no recuperado y eleva los costos de cobranza. Por otro lado, mantener un criterio de aprobación extremadamente rígido provoca el rechazo de clientes solventes, reduciendo los ingresos por intereses. La entidad financiera necesita optimizar la toma de decisiones para maximizar la aprobación de créditos rentables y controlar la tasa de morosidad.

---

### **Objetivo del Proyecto**
El objetivo principal de esta Prueba de Concepto (PoC) es desarrollar y evaluar un modelo predictivo (Regresión Logística) capaz de estimar la probabilidad de incumplimiento de pago (*default*) de los solicitantes de crédito a partir del archivo `application_train.csv`.

* **`TARGET = 0`**: Cliente que cumple correctamente sus pagos.
* **`TARGET = 1`**: Cliente que presenta incumplimiento o *default*.

---

### **Valor del Machine Learning para el Negocio**
La implementación de una solución basada en aprendizaje automático aporta un valor estratégico directo a la gestión del riesgo crediticio:

1. **Optimización de la Cartera y Rentabilidad:**
   * Minimiza el riesgo de morosidad al identificar patrones sutiles de incumplimiento.
   * Reduce el costo de oportunidad al evitar el rechazo injustificado de solicitantes aptos.

2. **Automatización y Consistencia:**
   * Permite evaluar solicitudes de forma estandarizada, rápida y sin sesgos subjetivos.

3. **Detección de Patrones Complejos:**
   * Identifica interacciones entre múltiples variables (socioeconómicas, demográficas y de scoring externo) que las reglas tradicionales basadas en umbrales fijos no capturan.

4. **Estrategia de Riesgo-Precio (*Risk-Based Pricing*):**
   * Al estimar la probabilidad continua de default de cada solicitante, la entidad puede graduar los montos de crédito, ajustar las tasas de interés o solicitar garantías adicionales según el perfil de riesgo del cliente.

---



### **Pregunta de Negocio**
> **¿Cómo podemos predecir con precisión la probabilidad de incumplimiento de pago (*default*) de los solicitantes de crédito a partir de su perfil socioeconómico y financiero, para optimizar las aprobaciones y reducir la tasa de morosidad sin impactar el crecimiento de la cartera?**

In [1]:
#Montar Google Drive en Colab
#from google.colab import drive
#drive.mount('/content/drive')

# Importar las librerías principales de Ciencia de Datos
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Configuración visual básica
%matplotlib inline
pd.set_option('display.max_columns', None)


In [2]:
"""import os
import pandas as pd
from IPython.display import display

# 1. Configurar el token de la API de Kaggle
os.environ['KAGGLE_API_TOKEN'] = "KGAT_beeea6ac31f6e1690b6f694660ebd8df"

# 2. Instalar la librería oficial de Kaggle y descargar el dataset
!pip install -q kaggle
!kaggle competitions download -c home-credit-default-risk

# 3. Descomprimir los archivos
!unzip -q home-credit-default-risk.zip
print("¡Descarga y descompresión completadas con éxito!")

# 4. Cargar todos los DataFrames
df_train = pd.read_csv('application_train.csv')
df_test = pd.read_csv('application_test.csv')
df_prev_app = pd.read_csv('previous_application.csv')
df_pos_cash = pd.read_csv('POS_CASH_balance.csv')
df_installments = pd.read_csv('installments_payments.csv')
df_credit_card = pd.read_csv('credit_card_balance.csv')
df_bureau = pd.read_csv('bureau.csv')
df_bureau_balance = pd.read_csv('bureau_balance.csv')
df_sub = pd.read_csv('sample_submission.csv')
df_columns_desc = pd.read_csv('HomeCredit_columns_description.csv', encoding='latin1')

# 5. Visualizar las primeras filas
print("--- Application Train ---")
display(df_train.head())
import pandas as pd
from IPython.display import display

# 1. Actualizar la librería de Kaggle para eliminar el Warning
!pip install --upgrade -q kaggle

# 2. Descargar el dataset de la competencia (requiere tener kaggle.json configurado)
!kaggle competitions download -c home-credit-default-risk

# 3. Descomprimir forzando el reemplazo silencioso (-o para overwrite, -q para quiet)
!unzip -o -q home-credit-default-risk.zip

# 4. Descargar el archivo auxiliar Excel desde GitHub
!curl -s -L -o Analisis_Previo.xlsx "https://raw.githubusercontent.com/nanomelk/Analista-de-Datos-II/main/Analisis_Previo.xlsx"

# 5. Cargar todos los DataFrames en memoria
print("\n--- Cargando Datasets en Pandas ---")
df_train = pd.read_csv('application_train.csv')
df_test = pd.read_csv('application_test.csv')

df_prev_app = pd.read_csv('previous_application.csv')
df_pos_cash = pd.read_csv('POS_CASH_balance.csv')
df_installments = pd.read_csv('installments_payments.csv')
df_credit_card = pd.read_csv('credit_card_balance.csv')

df_bureau = pd.read_csv('bureau.csv')
df_bureau_balance = pd.read_csv('bureau_balance.csv')

df_sub = pd.read_csv('sample_submission.csv')
df_columns_desc = pd.read_csv('HomeCredit_columns_description.csv', encoding='latin1')
df_analisis = pd.read_excel('Analisis_Previo.xlsx', header=4)

print("✅ Todos los archivos de Kaggle y GitHub han sido cargados con éxito.")

# 6. Muestra de los datos principales
display(df_train.head())"""
import os
import pandas as pd
from IPython.display import display

# 1. Configurar credenciales de Kaggle
# Reemplaza 'TU_USUARIO' y 'TU_KAGGLE_KEY' con las credenciales de tu archivo kaggle.json
os.environ['KAGGLE_API_TOKEN'] = "KGAT_beeea6ac31f6e1690b6f694660ebd8df"


# 2. Actualizar la herramienta de Kaggle
!pip install --upgrade -q kaggle

# 3. Descargar el dataset de la competencia desde Kaggle
print("--- Descargando datasets desde Kaggle ---")
!kaggle competitions download -c home-credit-default-risk

# 4. Descomprimir los archivos (-o reemplaza automáticamente, -q lo hace sin spam de texto)
print("--- Descomprimiendo archivos ---")
!unzip -o -q home-credit-default-risk.zip

# 5. Descargar el archivo auxiliar desde GitHub
print("--- Descargando Analisis_Previo.xlsx desde GitHub ---")
!curl -s -L -o Analisis_Previo.xlsx "https://raw.githubusercontent.com/nanomelk/Analista-de-Datos-II/main/Analisis_Previo.xlsx"

# 6. Cargar todos los DataFrames en Pandas
print("\n--- Cargando DataFrames en memoria ---")
df_train = pd.read_csv('application_train.csv')
df_test = pd.read_csv('application_test.csv')

df_prev_app = pd.read_csv('previous_application.csv')
df_pos_cash = pd.read_csv('POS_CASH_balance.csv')
df_installments = pd.read_csv('installments_payments.csv')
df_credit_card = pd.read_csv('credit_card_balance.csv')

df_bureau = pd.read_csv('bureau.csv')
df_bureau_balance = pd.read_csv('bureau_balance.csv')

df_sub = pd.read_csv('sample_submission.csv')
df_columns_desc = pd.read_csv('HomeCredit_columns_description.csv', encoding='latin1')
df_analisis = pd.read_excel('Analisis_Previo.xlsx', header=4)

print("✅ Todos los archivos cargados correctamente.")

# 7. Vista previa de los datos
display(df_train.head())

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.2/126.2 kB 5.4 MB/s eta 0:00:00
--- Descargando datasets desde Kaggle ---
100% 688M/688M [00:18<00:00, 39.3MB/s]

--- Descomprimiendo archivos ---
--- Descargando Analisis_Previo.xlsx desde GitHub ---

--- Cargando DataFrames en memoria ---
✅ Todos los archivos cargados correctamente.


,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_TYPE_SUITE,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,OCCUPATION_TYPE,CNT_FAM_MEMBERS,REGION_RATING_CLIENT,REGION_RATING_CLIENT_W_CITY,WEEKDAY_APPR_PROCESS_START,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,LIVE_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,LIVE_CITY_NOT_WORK_CITY,ORGANIZATION_TYPE,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,APARTMENTS_AVG,BASEMENTAREA_AVG,YEARS_BEGINEXPLUATATION_AVG,YEARS_BUILD_AVG,COMMONAREA_AVG,ELEVATORS_AVG,ENTRANCES_AVG,FLOORSMAX_AVG,FLOORSMIN_AVG,LANDAREA_AVG,LIVINGAPARTMENTS_AVG,LIVINGAREA_AVG,NONLIVINGAPARTMENTS_AVG,NONLIVINGAREA_AVG,APARTMENTS_MODE,BASEMENTAREA_MODE,YEARS_BEGINEXPLUATATION_MODE,YEARS_BUILD_MODE,COMMONAREA_MODE,ELEVATORS_MODE,ENTRANCES_MODE,FLOORSMAX_MODE,FLOORSMIN_MODE,LANDAREA_MODE,LIVINGAPARTMENTS_MODE,LIVINGAREA_MODE,NONLIVINGAPARTMENTS_MODE,NONLIVINGAREA_MODE,APARTMENTS_MEDI,BASEMENTAREA_MEDI,YEARS_BEGINEXPLUATATION_MEDI,YEARS_BUILD_MEDI,COMMONAREA_MEDI,ELEVATORS_MEDI,ENTRANCES_MEDI,FLOORSMAX_MEDI,FLOORSMIN_MEDI,LANDAREA_MEDI,LIVINGAPARTMENTS_MEDI,LIVINGAREA_MEDI,NONLIVINGAPARTMENTS_MEDI,NONLIVINGAREA_MEDI,FONDKAPREMONT_MODE,HOUSETYPE_MODE,TOTALAREA_MODE,WALLSMATERIAL_MODE,EMERGENCYSTATE_MODE,OBS_30_CNT_SOCIAL_CIRCLE,DEF_30_CNT_SOCIAL_CIRCLE,OBS_60_CNT_SOCIAL_CIRCLE,DEF_60_CNT_SOCIAL_CIRCLE,DAYS_LAST_PHONE_CHANGE,FLAG_DOCUMENT_2,FLAG_DOCUMENT_3,FLAG_DOCUMENT_4,FLAG_DOCUMENT_5,FLAG_DOCUMENT_6,FLAG_DOCUMENT_7,FLAG_DOCUMENT_8,FLAG_DOCUMENT_9,FLAG_DOCUMENT_10,FLAG_DOCUMENT_11,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,351000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.018801,-9461,-637,-3648.0,-2120,NaN,1,1,0,1,1,0,Laborers,1.0,2,2,WEDNESDAY,10,0,0,0,0,0,0,Business Entity Type 3,0.083037,0.262949,0.139376,0.0247,0.0369,0.9722,0.6192,0.0143,0.00,0.0690,0.0833,0.1250,0.0369,0.0202,0.0190,0.0000,0.0000,0.0252,0.0383,0.9722,0.6341,0.0144,0.0000,0.0690,0.0833,0.1250,0.0377,0.022,0.0198,0.0,0.0,0.0250,0.0369,0.9722,0.6243,0.0144,0.00,0.0690,0.0833,0.1250,0.0375,0.0205,0.0193,0.0000,0.00,reg oper account,block of flats,0.0149,"Stone, brick",No,2.0,2.0,2.0,2.0,-1134.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,1129500.0,Family,State servant,Higher education,Married,House / apartment,0.003541,-16765,-1188,-1186.0,-291,NaN,1,1,0,1,1,0,Core staff,2.0,1,1,MONDAY,11,0,0,0,0,0,0,School,0.311267,0.622246,NaN,0.0959,0.0529,0.9851,0.7960,0.0605,0.08,0.0345,0.2917,0.3333,0.0130,0.0773,0.0549,0.0039,0.0098,0.0924,0.0538,0.9851,0.8040,0.0497,0.0806,0.0345,0.2917,0.3333,0.0128,0.079,0.0554,0.0,0.0,0.0968,0.0529,0.9851,0.7987,0.0608,0.08,0.0345,0.2917,0.3333,0.0132,0.0787,0.0558,0.0039,0.01,reg oper account,block of flats,0.0714,Block,No,1.0,0.0,1.0,0.0,-828.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,135000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.010032,-19046,-225,-4260.0,-2531,26.0,1,1,1,1,1,0,Laborers,1.0,2,2,MONDAY,9,0,0,0,0,0,0,Government,NaN,0.555912,0.729567,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,

In [ ]:
"""import os
from google.colab import auth
from pydrive2.auth import GoogleAuth
from pydrive2.drive import GoogleDrive
from oauth2client.client import GoogleCredentials
import zipfile

# 1. Autenticar tu sesión de Google
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

# 2. Buscar el archivo zip dentro de tus compartidos
file_list = drive.ListFile({'q': "title = 'home-credit-default-risk.zip' and trashed = false"}).GetList()

if file_list:
    zip_id = file_list[0]['id']
    print(f"Descargando archivo ZIP (ID: {zip_id})...")

    zip_file = drive.CreateFile({'id': zip_id})
    zip_file.GetContentFile('home-credit-default-risk.zip')

    # 3. Descomprimir todos los CSVs en el directorio actual
    print("Descomprimiendo archivos...")
    with zipfile.ZipFile('home-credit-default-risk.zip', 'r') as zip_ref:
        zip_ref.extractall('.')

    print("¡Listo! Archivos descomprimidos:")
    print([f for f in os.listdir('.') if f.endswith('.csv')])
else:
    print("No se encontró el archivo zip en Compartidos conmigo.")"""

In [ ]:
#from google.colab import drive
#drive.mount('/content/drive')

In [3]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 286.2+ MB


In [4]:
# DETECCION DE NULOS
# ESTA FUNCIÓN CREA UN DATASET CON INFORMACIÓN DE CADA CAMPO, SIMILAR A INFO Y LO GUARDA EN UN ARCHIVO

def analizar_y_guardar_faltantes(df: pd.DataFrame, ruta_salida: str = "datos_faltantes.csv") -> pd.DataFrame:
    """
    Calcula la cantidad, porcentaje de datos faltantes y tipo de dato por columna
    en un DataFrame, y exporta el resultado a un archivo (CSV o Excel).
    """
    # 1. Obtenemos el conteo de nulos y el tipo de dato por columna
    nulos_por_columna = df.isna().sum()
    tipos_de_datos = df.dtypes

    # 2. Construimos el DataFrame de resultados
    df_resultado = pd.DataFrame({
        'columna': nulos_por_columna.index,
        'tipo_dato': tipos_de_datos.astype(str).values,
        'faltantes': nulos_por_columna.values,
        'porcentaje': (nulos_por_columna.values / len(df) * 100).round(2)
    })

    # 3. Guardar el archivo en formato CSV o Excel según la extensión
    if ruta_salida.endswith('.xlsx'):
        df_resultado.to_excel(ruta_salida, index=False)
    else:
        df_resultado.to_csv(ruta_salida, index=False)

    print(f"Reporte de faltantes guardado correctamente en: {ruta_salida}")

    return df_resultado

Volumen de Datos:

*   307,511 registros (filas que representan a cada solicitante de crédito)

*   122 características (columnas con atributos del cliente y la solicitud).




In [5]:
# visualizacion de los 3 primeros registros
df_train.head(3)

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_TYPE_SUITE,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,OCCUPATION_TYPE,CNT_FAM_MEMBERS,REGION_RATING_CLIENT,REGION_RATING_CLIENT_W_CITY,WEEKDAY_APPR_PROCESS_START,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,LIVE_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,LIVE_CITY_NOT_WORK_CITY,ORGANIZATION_TYPE,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,APARTMENTS_AVG,BASEMENTAREA_AVG,YEARS_BEGINEXPLUATATION_AVG,YEARS_BUILD_AVG,COMMONAREA_AVG,ELEVATORS_AVG,ENTRANCES_AVG,FLOORSMAX_AVG,FLOORSMIN_AVG,LANDAREA_AVG,LIVINGAPARTMENTS_AVG,LIVINGAREA_AVG,NONLIVINGAPARTMENTS_AVG,NONLIVINGAREA_AVG,APARTMENTS_MODE,BASEMENTAREA_MODE,YEARS_BEGINEXPLUATATION_MODE,YEARS_BUILD_MODE,COMMONAREA_MODE,ELEVATORS_MODE,ENTRANCES_MODE,FLOORSMAX_MODE,FLOORSMIN_MODE,LANDAREA_MODE,LIVINGAPARTMENTS_MODE,LIVINGAREA_MODE,NONLIVINGAPARTMENTS_MODE,NONLIVINGAREA_MODE,APARTMENTS_MEDI,BASEMENTAREA_MEDI,YEARS_BEGINEXPLUATATION_MEDI,YEARS_BUILD_MEDI,COMMONAREA_MEDI,ELEVATORS_MEDI,ENTRANCES_MEDI,FLOORSMAX_MEDI,FLOORSMIN_MEDI,LANDAREA_MEDI,LIVINGAPARTMENTS_MEDI,LIVINGAREA_MEDI,NONLIVINGAPARTMENTS_MEDI,NONLIVINGAREA_MEDI,FONDKAPREMONT_MODE,HOUSETYPE_MODE,TOTALAREA_MODE,WALLSMATERIAL_MODE,EMERGENCYSTATE_MODE,OBS_30_CNT_SOCIAL_CIRCLE,DEF_30_CNT_SOCIAL_CIRCLE,OBS_60_CNT_SOCIAL_CIRCLE,DEF_60_CNT_SOCIAL_CIRCLE,DAYS_LAST_PHONE_CHANGE,FLAG_DOCUMENT_2,FLAG_DOCUMENT_3,FLAG_DOCUMENT_4,FLAG_DOCUMENT_5,FLAG_DOCUMENT_6,FLAG_DOCUMENT_7,FLAG_DOCUMENT_8,FLAG_DOCUMENT_9,FLAG_DOCUMENT_10,FLAG_DOCUMENT_11,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,351000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.018801,-9461,-637,-3648.0,-2120,NaN,1,1,0,1,1,0,Laborers,1.0,2,2,WEDNESDAY,10,0,0,0,0,0,0,Business Entity Type 3,0.083037,0.262949,0.139376,0.0247,0.0369,0.9722,0.6192,0.0143,0.00,0.0690,0.0833,0.1250,0.0369,0.0202,0.0190,0.0000,0.0000,0.0252,0.0383,0.9722,0.6341,0.0144,0.0000,0.0690,0.0833,0.1250,0.0377,0.022,0.0198,0.0,0.0,0.0250,0.0369,0.9722,0.6243,0.0144,0.00,0.0690,0.0833,0.1250,0.0375,0.0205,0.0193,0.0000,0.00,reg oper account,block of flats,0.0149,"Stone, brick",No,2.0,2.0,2.0,2.0,-1134.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,1129500.0,Family,State servant,Higher education,Married,House / apartment,0.003541,-16765,-1188,-1186.0,-291,NaN,1,1,0,1,1,0,Core staff,2.0,1,1,MONDAY,11,0,0,0,0,0,0,School,0.311267,0.622246,NaN,0.0959,0.0529,0.9851,0.7960,0.0605,0.08,0.0345,0.2917,0.3333,0.0130,0.0773,0.0549,0.0039,0.0098,0.0924,0.0538,0.9851,0.8040,0.0497,0.0806,0.0345,0.2917,0.3333,0.0128,0.079,0.0554,0.0,0.0,0.0968,0.0529,0.9851,0.7987,0.0608,0.08,0.0345,0.2917,0.3333,0.0132,0.0787,0.0558,0.0039,0.01,reg oper account,block of flats,0.0714,Block,No,1.0,0.0,1.0,0.0,-828.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,135000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.010032,-19046,-225,-4260.0,-2531,26.0,1,1,1,1,1,0,Laborers,1.0,2,2,MONDAY,9,0,0,0,0,0,0,Government,NaN,0.555912,0.729567,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,

In [6]:
# visualizacion de datos estadisticos de las variables
df_train.describe()

,SK_ID_CURR,TARGET,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,REGION_RATING_CLIENT,REGION_RATING_CLIENT_W_CITY,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,LIVE_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,LIVE_CITY_NOT_WORK_CITY,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,APARTMENTS_AVG,BASEMENTAREA_AVG,YEARS_BEGINEXPLUATATION_AVG,YEARS_BUILD_AVG,COMMONAREA_AVG,ELEVATORS_AVG,ENTRANCES_AVG,FLOORSMAX_AVG,FLOORSMIN_AVG,LANDAREA_AVG,LIVINGAPARTMENTS_AVG,LIVINGAREA_AVG,NONLIVINGAPARTMENTS_AVG,NONLIVINGAREA_AVG,APARTMENTS_MODE,BASEMENTAREA_MODE,YEARS_BEGINEXPLUATATION_MODE,YEARS_BUILD_MODE,COMMONAREA_MODE,ELEVATORS_MODE,ENTRANCES_MODE,FLOORSMAX_MODE,FLOORSMIN_MODE,LANDAREA_MODE,LIVINGAPARTMENTS_MODE,LIVINGAREA_MODE,NONLIVINGAPARTMENTS_MODE,NONLIVINGAREA_MODE,APARTMENTS_MEDI,BASEMENTAREA_MEDI,YEARS_BEGINEXPLUATATION_MEDI,YEARS_BUILD_MEDI,COMMONAREA_MEDI,ELEVATORS_MEDI,ENTRANCES_MEDI,FLOORSMAX_MEDI,FLOORSMIN_MEDI,LANDAREA_MEDI,LIVINGAPARTMENTS_MEDI,LIVINGAREA_MEDI,NONLIVINGAPARTMENTS_MEDI,NONLIVINGAREA_MEDI,TOTALAREA_MODE,OBS_30_CNT_SOCIAL_CIRCLE,DEF_30_CNT_SOCIAL_CIRCLE,OBS_60_CNT_SOCIAL_CIRCLE,DEF_60_CNT_SOCIAL_CIRCLE,DAYS_LAST_PHONE_CHANGE,FLAG_DOCUMENT_2,FLAG_DOCUMENT_3,FLAG_DOCUMENT_4,FLAG_DOCUMENT_5,FLAG_DOCUMENT_6,FLAG_DOCUMENT_7,FLAG_DOCUMENT_8,FLAG_DOCUMENT_9,FLAG_DOCUMENT_10,FLAG_DOCUMENT_11,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
count,307511.000000,307511.000000,307511.000000,3.075110e+05,3.075110e+05,307499.000000,3.072330e+05,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,104582.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307509.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,134133.000000,3.068510e+05,246546.000000,151450.00000,127568.000000,157504.000000,103023.000000,92646.000000,143620.000000,152683.000000,154491.000000,98869.000000,124921.000000,97312.000000,153161.000000,93997.000000,137829.000000,151450.000000,127568.000000,157504.000000,103023.000000,92646.000000,143620.000000,152683.000000,154491.000000,98869.000000,124921.000000,97312.000000,153161.000000,93997.000000,137829.000000,151450.000000,127568.000000,157504.000000,103023.000000,92646.000000,143620.000000,152683.000000,154491.000000,98869.000000,124921.000000,97312.000000,153161.000000,93997.000000,137829.000000,159080.000000,306490.000000,306490.000000,306490.000000,306490.000000,307510.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.00000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,307511.000000,265992.000000,265992.000000,265992.000000,265992.000000,265992.000000,265992.000000
mean,278180.518577,0.080729,0.417052,1.687979e+05,5.990260e+05,27108.573909,5.383962e+05,0.020868,-16036.995067,63815.045904,-4986.120328,-2994.202373,12.061091,0.999997,0.819889,0.199368,0.998133,0.281066,0.056720,2.152665,2.052463,2.031521,12.063419,0.015144,0.050769,0.040659,0.078173,0.230454,0.179555,0.502130,5.143927e-01,0.510853,0.11744,0.088442,0.977735,0.752471,0.044621,0.078942,0.149725,0.226282,0.231894,0.066333,0.100775,0.107399,0.008809,0.028358,0.114231,0.087543,0.977065,0.759637,0.042553,0.074490,0.145193,0.222315,0.228058,0.064958,0.105645,0.105975,0.008076,0.027022,0.117850,0.087955,0.97

In [7]:
# Conteo del tipo de dato agrupado en columnas
conteo_tipos = df_train.dtypes.value_counts().reset_index()
conteo_tipos.columns = ['Tipo de Dato', 'Cantidad de Columnas']

display(conteo_tipos)

,Tipo de Dato,Cantidad de Columnas
0,float64,65
1,int64,41
2,object,16


Con el objetivo de comprender el dataset, se visualizan
 las variables con tipo de dato "object"

In [ ]:
# Filtrar solo las columnas categóricas (objeto)
cols_object = df_train.select_dtypes(include=['object'])

# Ver la cantidad de columnas objeto y sus primeras filas
print(f"Total de columnas categóricas: {cols_object.shape[1]}\n")
display(cols_object.head())

In [ ]:
# Analisis de los valores únicos que posee cada columna categórica
df_train.select_dtypes(include=['object']).nunique()

### Valores Nulos del dataset

In [ ]:
# 1. Calcular nulos totales y su porcentaje por columna
valores_nulos_totales = df_train.isnull().sum()
porcentaje_nulos = (valores_nulos_totales / len(df_train)) * 100

# 2. Unir ambos datos en un DataFrame ordenado
tabla_nulos = pd.DataFrame({
    'Valores Nulos': valores_nulos_totales,
    'Porcentaje (%)': porcentaje_nulos.round(2)
})

# 3. Filtrar solo las columnas que SI tienen nulos y ordenar de mayor a menor
tabla_nulos = tabla_nulos[tabla_nulos['Valores Nulos'] > 0].sort_values(by='Porcentaje (%)', ascending=False)

print(f"Total de columnas con valores nulos: {len(tabla_nulos)}")
display(tabla_nulos)

## **Criterio de Tratamiento para Variables con Alto Porcentaje de Datos Faltantes**

Al analizar la distribución de los valores nulos en el dataset, se observa que una gran cantidad de columnas superan el 40% y 50% de datos ausentes (alcanzando en varios casos casi un 70%).

Ante un volumen de datos faltantes tan elevado, la estrategia metodológica que tomamos para tratar dichas variables es el de la eliminacion de las mismas en lugar de aplicar técnicas de imputación (relleno por media o mediana) por los siguientes motivos técnicos y estadísticos:


*   **Introducción de Sesgos Severos:** Imputar artificialmente más del 40% de los registros en una columna altera de forma sustancial la varianza, la desviación estándar y la distribución original de la variable, distorsionando la realidad de los datos.

*   **Inyección de Ruido al Modelo:** Rellenar valores masivamente con supuestos genera patrones ficticios que no aportan información predictiva real y pueden perjudicar el desempeño y la capacidad de generalización de los algoritmos de Machine Learning.


*   **Pérdida de Representatividad:** Cuando la ausencia de información es tan masiva, la variable pierde su utilidad analítica para caracterizar la población de clientes.


### **Estrategia a aplicar:**

Se mantendrán únicamente aquellas variables cuyo porcentaje de nulos sea inferior a un umbral operativo (ej. < 50%), reservando la imputación estadística exclusivamente para variables clave con porcentajes de ausencia marginales (< 5%).

In [ ]:
#  Identificar las columnas a conservar(menos del 50% de nulos)
columnas_validas = tabla_nulos[tabla_nulos['Porcentaje (%)'] < 50].index

# Incluir las columnas que no tenían ningún nulo
columnas_sin_nulos = df_train.columns.difference(tabla_nulos.index)
todas_las_columnas_a_conservar = columnas_sin_nulos.union(columnas_validas)

#  Crear un nuevo dataset reducido por columnas (Se crea un nuevo Dataframe 'df_reducido' para trabajar la transformación.
#  Esto evita la mutación de datos y conserva 'df_train' como fuente intacta)
df_reducido = df_train[todas_las_columnas_a_conservar].copy()

#  Comprobar dimensiones
print(f"Dataset original (df_train): {df_train.shape[1]} columnas")
print(f"Nuevo dataset (df_reducido): {df_reducido.shape[1]} columnas")
print(f"Columnas descartadas por exceso de nulos: {df_train.shape[1] - df_reducido.shape[1]}")

# **Analisis de las 41 columnas con una cantidad de datos nulos superior al 50%**

In [ ]:
#Identificar las 41 columnas con más del 50% de nulos
porcentaje_nulos = (df_train.isnull().sum() / len(df_train)) * 100
columnas_eliminadas = porcentaje_nulos[porcentaje_nulos > 50].index.tolist()

#Crear el dataset reducido (81 columnas)
df_reducido = df_train.drop(columns=columnas_eliminadas).copy()

#Mostrar la lista exacta de las columnas eliminadas
print(f"Total de columnas eliminadas: {len(columnas_eliminadas)}\n")
print(columnas_eliminadas)


### **Listado Detallado de las 41 Columnas Eliminadas (> 50% de Nulos)**

Se descartaron exactamente **41 variables** por superar el 50% de datos faltantes para evitar sesgar el análisis mediante imputaciones artificiales:

| # | Nombre de la Columna | Categoría | Descripción |
| :-: | :--- | :--- | :--- |
| **1** | `OWN_CAR_AGE` | Vehículo | Antigüedad (en años) del vehículo del cliente. |
| **2** | `EXT_SOURCE_1` | Scoring Externo | Puntuación normalizada proveniente de la fuente externa #1. |
| **3** | `APARTMENTS_AVG` | Vivienda (Promedio) | Tamaño del departamento normalizado. |
| **4** | `APARTMENTS_MODE` | Vivienda (Moda) | Tamaño del departamento normalizado. |
| **5** | `APARTMENTS_MEDI` | Vivienda (Mediana) | Tamaño del departamento normalizado. |
| **6** | `BASEMENTAREA_AVG` | Vivienda (Promedio) | Área del sótano normalizada. |
| **7** | `BASEMENTAREA_MODE` | Vivienda (Moda) | Área del sótano normalizada. |
| **8** | `BASEMENTAREA_MEDI` | Vivienda (Mediana) | Área del sótano normalizada. |
| **9** | `YEARS_BUILD_AVG` | Vivienda (Promedio) | Antigüedad / año de construcción del edificio. |
| **10** | `YEARS_BUILD_MODE` | Vivienda (Moda) | Antigüedad / año de construcción del edificio. |
| **11** | `YEARS_BUILD_MEDI` | Vivienda (Mediana) | Antigüedad / año de construcción del edificio. |
| **12** | `COMMONAREA_AVG` | Vivienda (Promedio) | Área común del inmueble normalizada. |
| **13** | `COMMONAREA_MODE` | Vivienda (Moda) | Área común del inmueble normalizada. |
| **14** | `COMMONAREA_MEDI` | Vivienda (Mediana) | Área común del inmueble normalizada. |
| **15** | `ELEVATORS_AVG` | Vivienda (Promedio) | Cantidad de ascensores en el edificio. |
| **16** | `ELEVATORS_MODE` | Vivienda (Moda) | Cantidad de ascensores en el edificio. |
| **17** | `ELEVATORS_MEDI` | Vivienda (Mediana) | Cantidad de ascensores en el edificio. |
| **18** | `ENTRANCES_AVG` | Vivienda (Promedio) | Cantidad de entradas al edificio. |
| **19** | `ENTRANCES_MODE` | Vivienda (Moda) | Cantidad de entradas al edificio. |
| **20** | `ENTRANCES_MEDI` | Vivienda (Mediana) | Cantidad de entradas al edificio. |
| **21** | `FLOORSMAX_AVG` | Vivienda (Promedio) | Número máximo de pisos en el edificio. |
| **22** | `FLOORSMIN_AVG` | Vivienda (Promedio) | Número mínimo de pisos en el edificio. |
| **23** | `FLOORSMIN_MODE` | Vivienda (Moda) | Número mínimo de pisos en el edificio. |
| **24** | `FLOORSMIN_MEDI` | Vivienda (Mediana) | Número mínimo de pisos en el edificio. |
| **25** | `LANDAREA_AVG` | Vivienda (Promedio) | Área del terreno normalizada. |
| **26** | `LANDAREA_MODE` | Vivienda (Moda) | Área del terreno normalizada. |
| **27** | `LANDAREA_MEDI` | Vivienda (Mediana) | Área del terreno normalizada. |
| **28** | `LIVINGAPARTMENTS_AVG` | Vivienda (Promedio) | Área departamental habitable normalizada. |
| **29** | `LIVINGAPARTMENTS_MODE` | Vivienda (Moda) | Área departamental habitable normalizada. |
| **30** | `LIVINGAPARTMENTS_MEDI` | Vivienda (Mediana) | Área departamental habitable normalizada. |
| **31** | `LIVINGAREA_AVG` | Vivienda (Promedio) | Área habitable total normalizada. |
| **32** | `LIVINGAREA_MODE` | Vivienda (Moda) | Área habitable total normalizada. |
| **33** | `LIVINGAREA_MEDI` | Vivienda (Mediana) | Área habitable total normalizada. |
| **34** | `NONLIVINGAPARTMENTS_AVG` | Vivienda (Promedio) | Área departamental no habitable normalizada. |
| **35** | `NONLIVINGAPARTMENTS_MODE` | Vivienda (Moda) | Área departamental no habitable normalizada. |
| **36** | `NONLIVINGAPARTMENTS_MEDI` | Vivienda (Mediana) | Área departamental no habitable normalizada. |
| **37** | `NONLIVINGAREA_AVG` | Vivienda (Promedio) | Área no habitable total normalizada. |
| **38** | `NONLIVINGAREA_MODE` | Vivienda (Moda) | Área no habitable total normalizada. |
| **39** | `NONLIVINGAREA_MEDI` | Vivienda (Mediana) | Área no habitable total normalizada. |
| **40** | `HOUSETYPE_MODE` | Vivienda (Moda) | Tipo de vivienda u hogar. |
| **41** | `WALLSMATERIAL_MODE` | Vivienda (Moda) | Material de construcción de las paredes externas. |

---

> **Conclusión técnica:**
> * **39 variables de infraestructura:** Corresponden a detalles arquitectónicos de la vivienda que la mayoría de los clientes no declararon.
> * **`OWN_CAR_AGE`:** Presente únicamente en clientes que declararon poseer vehículo propio.
> * **`EXT_SOURCE_1`:** Score crediticio externo ausente en más del 50% de la muestra.

# **Estrategia de Categorización y Tratamiento de Valores Nulos Restantes**

Tras haber removido las columnas con ausencia masiva (> 50%), el dataset `df_reducido` aún presenta variables con datos faltantes. Debido a que la cantidad de nulos difiere sustancialmente entre columnas, **no resulta conveniente aplicar una única técnica de imputación global**, ya que esto podría sesgar la distribución de los datos o distorsionar patrones predictivos clave.

Para abordar este proceso de manera estratégica y metodológicamente rigurosa, se clasificaron las variables con nulos en tres franjas según su porcentaje de ausencia:

* **1. Nivel Alto (30% a 50%):** Reemplazar las ausencias por la etiqueta `Desconocido` en variables categóricas. En variables numéricas de este rango VER QUE HACER (ANALIZAR CASOS AISLADOS).
* **2. Nivel Medio (5% a 30%):** Se aplica imputación por **mediana**  para preservar el alto poder predictivo de estas métricas.
* **3. Nivel Bajo (< 5%):**  Al representar una pérdida de información mínima, el riesgo de distorsión estadística es bajo. Se aplica imputación directa por **mediana** para variables continuas y por la **moda** (valor más frecuente) para variables categóricas.

Complementariamente al criterio cuantitativo aplicado por franjas de vacíos, se llevará a cabo un análisis cualitativo e individual por variable. El objetivo es evaluar su relevancia del negocio y capacidad predictiva respecto al fenómeno en estudio. De este modo, la decisión final de inclusión o descarte no dependerá únicamente del volumen de faltantes; variables con baja tasa de ausencias podrían ser omitidas por redundancia o falta de poder explicativo, mientras que variables pertenecientes a franjas de alta vaciedad podrían ser retenidas e imputadas mediante técnicas avanzadas si aportan valor crítico al modelo.


---


**Objetivo del Enfoque:**
Esta segmentación por niveles asegura un tratamiento diferenciado y adaptado al volumen de datos faltantes, preservando la calidad estadística del dataset antes de proceder con el modelado.

In [ ]:
#Calcular nulos y porcentajes únicamente sobre df_reducido
nulos = df_reducido.isnull().sum()
porcentajes = (nulos / len(df_reducido)) * 100

#Filtrar solo las columnas que conservan nulos (> 0%)
df_nulos = pd.DataFrame(
    {
        'Columna': nulos.index,
        'Tipo_Dato': df_reducido.dtypes.values,
        'Valores_Nulos': nulos.values,
        'Porcentaje (%)': porcentajes.round(2).values,
    }
)
df_nulos = df_nulos[df_nulos['Valores_Nulos'] > 0]


#Función para clasificar en las 3 franjas de criticidad
def clasificar_franja(pct):
    if pct >= 30:
        return '1. Alto (30% - 50%)'
    elif pct >= 5:
        return '2. Medio (5% - 30%)'
    else:
        return '3. Bajo (< 5%)'


df_nulos['Franja'] = df_nulos['Porcentaje (%)'].apply(clasificar_franja)

#Ordenar y mostrar la tabla agrupada
tabla_clasificada = df_nulos.sort_values(
    by=['Franja', 'Porcentaje (%)'], ascending=[True, False]
)

display(tabla_clasificada.reset_index(drop=True))



### **Resumen General de la Distribución de Variables (122 Columnas Totales)**

El total de **122 variables** del dataset original se clasificó de la siguiente manera:

* **Columnas Eliminadas (41 variables):** Superaban el 50% de valores faltantes.
* **Dataset Reducido / Conservado (81 variables):**
  * **Sin nulos (55 columnas):** Tienen el **100% de sus datos completos**.
  * **Con nulos (26 columnas):** Variables que conservan faltantes y serán imputadas:
    * **Franja Alta (30% - 50%):** 9 columnas
    * **Franja Media (5% - 30%):** 7 columnas
    * **Franja Baja (< 5%):** 10 columnas

---

### **Desglose de las 26 Variables a Imputar por Franja**

#### **1. Franja Alta (30% a 50% de Nulos - 9 columnas)**

| Nombre de la Columna | Tipo de Dato | % Nulos | Descripción |
| :--- | :--- | :-: | :--- |
| **`FLOORSMAX_AVG`** | Numérico | 49.76% | Número máximo de pisos en el edificio (promedio). |
| **`FLOORSMAX_MODE`** | Numérico | 49.76% | Número máximo de pisos en el edificio (moda). |
| **`FLOORSMAX_MEDI`** | Numérico | 49.76% | Número máximo de pisos en el edificio (mediana). |
| **`YEARS_BEGINEXPLUATATION_AVG`** | Numérico | 48.78% | Antigüedad/inicio de explotación del edificio (promedio). |
| **`YEARS_BEGINEXPLUATATION_MODE`** | Numérico | 48.78% | Antigüedad/inicio de explotación del edificio (moda). |
| **`YEARS_BEGINEXPLUATATION_MEDI`** | Numérico | 48.78% | Antigüedad/inicio de explotación del edificio (mediana). |
| **`TOTALAREA_MODE`** | Numérico | 48.27% | Área total del inmueble normalizada (moda). |
| **`EMERGENCYSTATE_MODE`** | Categórico | 47.40% | Indicador si la vivienda presenta estado de emergencia/riesgo. |
| **`OCCUPATION_TYPE`** | Categórico | 31.35% | Tipo de ocupación, profesión u oficio que desempeña el cliente. |

---

#### **2. Franja Media (5% a 30% de Nulos - 7 columnas)**

| Nombre de la Columna | Tipo de Dato | % Nulos | Descripción |
| :--- | :--- | :-: | :--- |
| **`EXT_SOURCE_3`** | Numérico | 19.83% | Puntuación/score de riesgo normalizado de la fuente externa #3. |
| **`AMT_REQ_CREDIT_BUREAU_HOUR`** | Numérico | 13.50% | Consultas al Bureau de Crédito 1 hora antes de la solicitud. |
| **`AMT_REQ_CREDIT_BUREAU_DAY`** | Numérico | 13.50% | Consultas al Bureau de Crédito 1 día antes de la solicitud. |
| **`AMT_REQ_CREDIT_BUREAU_WEEK`** | Numérico | 13.50% | Consultas al Bureau de Crédito 1 semana antes de la solicitud. |
| **`AMT_REQ_CREDIT_BUREAU_MON`** | Numérico | 13.50% | Consultas al Bureau de Crédito 1 mes antes de la solicitud. |
| **`AMT_REQ_CREDIT_BUREAU_QRT`** | Numérico | 13.50% | Consultas al Bureau de Crédito 3 meses antes de la solicitud. |
| **`AMT_REQ_CREDIT_BUREAU_YEAR`** | Numérico | 13.50% | Consultas al Bureau de Crédito 1 año antes de la solicitud. |

---

#### **3. Franja Baja (< 5% de Nulos - 10 columnas)**

| Nombre de la Columna | Tipo de Dato | % Nulos | Descripción |
| :--- | :--- | :-: | :--- |
| **`NAME_TYPE_SUITE`** | Categórico | 0.42% | Acompañante presente al solicitar el préstamo. |
| **`OBS_30_CNT_SOCIAL_CIRCLE`** | Numérico | 0.33% | Personas del entorno social observadas con atrasos de 30 días. |
| **`DEF_30_CNT_SOCIAL_CIRCLE`** | Numérico | 0.33% | Personas del entorno social que cayeron en mora a 30 días. |
| **`OBS_60_CNT_SOCIAL_CIRCLE`** | Numérico | 0.33% | Personas del entorno social observadas con atrasos de 60 días. |
| **`DEF_60_CNT_SOCIAL_CIRCLE`** | Numérico | 0.33% | Personas del entorno social que cayeron en mora a 60 días. |
| **`EXT_SOURCE_2`** | Numérico | 0.21% | Puntuación/score de riesgo normalizado de la fuente externa #2. |
| **`AMT_GOODS_PRICE`** | Numérico | 0.09% | Precio del bien o servicio para el cual se otorga el crédito. |
| **`AMT_ANNUITY`** | Numérico | <0.01% | Monto de la cuota/anualidad fija a pagar por el préstamo. |
| **`CNT_FAM_MEMBERS`** | Numérico | <0.01% | Cantidad total de integrantes dentro del grupo familiar. |
| **`DAYS_LAST_PHONE_CHANGE`** | Numérico | <0.01% | Días transcurridos desde que el cliente cambió su teléfono. |

En el análisis inicial del dataset, se recomienda descartar un total de 80 variables de las 122 disponibles de entrada
. La eliminación de estas columnas se fundamenta en criterios técnicos como la falta de variabilidad, redundancia severa (multicolinealidad) o exceso de valores faltantes, lo cual optimiza el rendimiento de los modelos predictivos

In [ ]:
# ==============================================================================
# VARIABLES PRIORITARIAS PARA EL ANÁLISIS EXPLORATORIO DE DATOS (EDA)
# ==============================================================================
"""
A continuación se detallan las variables clave del archivo 'application_train.csv'
que se conservarán para la Evidencia 1, estructuradas según el tipo de información.

+------------------------+---------------------+-----------------------------------------------+---------------------------------------------------------------------------------------------------------+
| Tipo Variable          | Variable            | Descripción                                   | Fundamento                                                                                              |
+------------------------+---------------------+-----------------------------------------------+---------------------------------------------------------------------------------------------------------+
| Variable Objetivo      | TARGET              | 1 = Cliente en default, 0 = Paga bien         | Variable central de predicción. Presenta fuerte desbalance (~92% vs 8%) que fundamenta el uso de AUC-ROC. [2, 3]
| Variables Financieras  | AMT_INCOME_TOTAL    | Ingresos totales declarados del cliente       | Mide la capacidad económica básica para hacer frente al préstamo solicitado. [4]                       |
| Variables Financieras  | AMT_CREDIT          | Monto neto del préstamo otorgado              | Determina la magnitud del riesgo de crédito asumido por la financiera. [4]                             |
| Variables Financieras  | AMT_ANNUITY         | Cuota mensual de amortización del crédito     | Esencial para evaluar la tasa de esfuerzo de pago sobre los ingresos declarados. [5]                   |
| Variables Financieras  | AMT_GOODS_PRICE     | Precio del bien para el que se pide el crédito| Relación entre el valor del bien de consumo y el préstamo otorgado (LTV). [5]                         |
| Variables Demográficas | CODE_GENDER         | Género del solicitante                        | Segmentación básica y detección de sesgos o patrones de comportamiento demográfico. [5]                 |
| Variables Demográficas | NAME_EDUCATION_TYPE | Máximo nivel académico alcanzado              | Se correlaciona con la estabilidad laboral, formalidad de ingresos y menor riesgo histórico. [5]       |
| Variables Demográficas | NAME_FAMILY_STATUS  | Estado civil                                  | Informa sobre la estructura de responsabilidades del hogar y gastos compartidos. [5]                    |
| Variables Demográficas | NAME_INCOME_TYPE    | Actividad económica que genera ingresos       | Determina la regularidad y predictibilidad de los ingresos para el pago mensual. [5]                    |
| Variables Demográficas | NAME_HOUSING_TYPE   | Tipo de vivienda (alquiler, propia, etc.)     | Identifica costos de vivienda implícitos y el nivel de arraigo residencial. [5]                        |
| Variables Demográficas | CNT_CHILDREN        | Cantidad de hijos declarados                  | Carga financiera del hogar. Requiere limpieza previa por outliers con errores de carga (ej. 20 hijos). [5, 6]
| Variables Demográficas | CNT_FAM_MEMBERS     | Total de miembros del grupo familiar          | Permite dimensionar la base mínima de egresos necesarios para el sustento del hogar. [5]               |
| Tiempos Relativos      | DAYS_BIRTH          | Edad en días al momento de la solicitud       | Predictor de madurez de pago. Se debe transformar a años durante el EDA. [5]                           |
| Tiempos Relativos      | DAYS_EMPLOYED       | Antigüedad laboral medida en días             | Indicador de estabilidad económica. Requiere tratar el valor anómalo de 365243 (jubilados/desempleo). [5, 7]|
| Tiempos Relativos      | DAYS_REGISTRATION   | Días desde el último cambio de dirección      | Proxy indirecto de estabilidad geográfica y arraigo local. [5]                                         |
| Tiempos Relativos      | DAYS_ID_PUBLISH     | Días desde la renovación del documento        | Control operacional; valida que la identificación física no esté desactualizada. [5]                  |
| Scores Externos        | EXT_SOURCE_1        | Score crediticio de fuente externa 1          | Altísima correlación con default. Aunque sea "caja negra", es crítico para el modelo predictivo. [5, 7]|
| Scores Externos        | EXT_SOURCE_2        | Score crediticio de fuente externa 2          | Fuente de buró externa con mayor poder discriminante de mora en este dataset. [5]                      |
| Scores Externos        | EXT_SOURCE_3        | Score crediticio de fuente externa 3          | Complementa la terna de scores externos, mejorando sustancialmente el baseline. [5]                    |
| Validación y Calidad   | FLAG_DOCUMENT_3     | Entregó el documento principal 3              | Único indicador de entrega de documentación útil con variabilidad estadística relevante. [8]           |
| Validación y Calidad   | OCCUPATION_TYPE     | Tipo de ocupación laboral específica          | Útil para micro-segmentar riesgo profesional, a pesar de tener >30% de nulos. [6]                      |
| Validación y Calidad   | NAME_CONTRACT_TYPE  | Tipo de préstamo (efectivo o rotativo)        | Identifica el tipo de amortización aplicable a la solicitud. [1, 9]                                   |
| Validación y Calidad   | FLAG_OWN_CAR        | Indica si el cliente es dueño de auto          | Proxy indirecto de patrimonio y activos de mediano plazo. [1]                                           |
| Validación y Calidad   | FLAG_OWN_REALTY     | Indica si posee propiedades inmuebles         | Refleja solvencia a largo plazo y un mayor compromiso de cumplimiento. [1]                              |
+------------------------+---------------------+-----------------------------------------------+---------------------------------------------------------------------------------------------------------+
"""

# SELECCIÓN PREVIA DE VARIABLES UTILES

In [8]:
# Leer archivo excel denominado Analisis_Previo.xlsx que contiene la información de las variables a analizar y una breve descripción de cada una de ellas
# Los encabezados arrancan en la fila 4
#df_analisis = pd.read_excel('Analisis_Previo.xlsx', header=4)
import pandas as pd
from IPython.display import display

# 1. URL raw del archivo en tu GitHub
url = "https://raw.githubusercontent.com/nanomelk/Analista-de-Datos-II/main/Analisis_Previo.xlsx"

# 2. Descargar el archivo al entorno local de Colab usando curl
!curl -s -L -o Analisis_Previo.xlsx "{url}"

# 3. Leer el Excel con Pandas (aplicando header=4 como en tu script original)
df_analisis = pd.read_excel('Analisis_Previo.xlsx', header=4)

print("✅ Archivo cargado exitosamente desde GitHub:")
display(df_analisis.head())

✅ Archivo cargado exitosamente desde GitHub:


,Tipo Variable,Variable,Descripción,Fundamento
0,Variable Objetivo,TARGET,Identificador de si el cliente presentó dificu...,Es la variable central de clasificación. Su an...
1,Variables Financieras,AMT_INCOME_TOTAL,Ingresos declarados totales del solicitante.,"Mide la capacidad de generación de ingresos, e..."
2,Variables Financieras,AMT_CREDIT,Monto total de financiamiento otorgado en el p...,Determina el tamaño del riesgo de crédito asum...
3,Variables Financieras,AMT_ANNUITY,Cuota periódica (anualidad/mensualidad) de amo...,Indispensable para calcular ratios de endeudam...
4,Variables Financieras,AMT_GOODS_PRICE,Precio comercial del bien de consumo para el c...,Permite conocer el porcentaje de financiamient...


In [9]:
df_analisis.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24 entries, 0 to 23
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Tipo Variable  24 non-null     object
 1   Variable       24 non-null     object
 2   Descripción    24 non-null     object
 3   Fundamento     24 non-null     object
dtypes: object(4)
memory usage: 900.0+ bytes


In [10]:
# Crear una lista con las columnas que se conservarán para el análisis final
columnas_a_conservar = df_analisis['Variable'].tolist()

In [11]:
# Cargar el dataset (application_train.csv) con las columnas seleccionadas para el analisis
df = pd.read_csv('application_train.csv', usecols=columnas_a_conservar)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Data columns (total 24 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   TARGET               307511 non-null  int64  
 1   NAME_CONTRACT_TYPE   307511 non-null  object 
 2   CODE_GENDER          307511 non-null  object 
 3   FLAG_OWN_CAR         307511 non-null  object 
 4   FLAG_OWN_REALTY      307511 non-null  object 
 5   CNT_CHILDREN         307511 non-null  int64  
 6   AMT_INCOME_TOTAL     307511 non-null  float64
 7   AMT_CREDIT           307511 non-null  float64
 8   AMT_ANNUITY          307499 non-null  float64
 9   AMT_GOODS_PRICE      307233 non-null  float64
 10  NAME_INCOME_TYPE     307511 non-null  object 
 11  NAME_EDUCATION_TYPE  307511 non-null  object 
 12  NAME_FAMILY_STATUS   307511 non-null  object 
 13  NAME_HOUSING_TYPE    307511 non-null  object 
 14  DAYS_BIRTH           307511 non-null  int64  
 15  DAYS_EMPLOYED    

In [13]:
# identificar las columnas que tienen valores nulos en el dataset final
reporte_faltantes = analizar_y_guardar_faltantes(df, ruta_salida='reporte_faltantes_final.xlsx')

Reporte de faltantes guardado correctamente en: reporte_faltantes_final.xlsx


In [14]:
# Visualizar el reporte de valores faltantes, para variables con porcentaje de nulos mayor a 0%
reporte_faltantes[reporte_faltantes['porcentaje'] > 0]

,columna,tipo_dato,faltantes,porcentaje
9,AMT_GOODS_PRICE,float64,278,0.09
18,OCCUPATION_TYPE,object,96391,31.35
20,EXT_SOURCE_1,float64,173378,56.38
21,EXT_SOURCE_2,float64,660,0.21
22,EXT_SOURCE_3,float64,60965,19.83


In [15]:
# DESCRIBE DE LOS CAMPOS DE INTERES NUMERICOS
df.describe()

,TARGET,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,CNT_FAM_MEMBERS,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,FLAG_DOCUMENT_3
count,307511.000000,307511.000000,3.075110e+05,3.075110e+05,307499.000000,3.072330e+05,307511.000000,307511.000000,307511.000000,307511.000000,307509.000000,134133.000000,3.068510e+05,246546.000000,307511.000000
mean,0.080729,0.417052,1.687979e+05,5.990260e+05,27108.573909,5.383962e+05,-16036.995067,63815.045904,-4986.120328,-2994.202373,2.152665,0.502130,5.143927e-01,0.510853,0.710023
std,0.272419,0.722121,2.371231e+05,4.024908e+05,14493.737315,3.694465e+05,4363.988632,141275.766519,3522.886321,1509.450419,0.910682,0.211062,1.910602e-01,0.194844,0.453752
min,0.000000,0.000000,2.565000e+04,4.500000e+04,1615.500000,4.050000e+04,-25229.000000,-17912.000000,-24672.000000,-7197.000000,1.000000,0.014568,8.173617e-08,0.000527,0.000000
25%,0.000000,0.000000,1.125000e+05,2.700000e+05,16524.000000,2.385000e+05,-19682.000000,-2760.000000,-7479.500000,-4299.000000,2.000000,0.334007,3.924574e-01,0.370650,0.000000
50%,0.000000,0.000000,1.471500e+05,5.135310e+05,24903.000000,4.500000e+05,-15750.000000,-1213.000000,-4504.000000,-3254.000000,2.000000,0.505998,5.659614e-01,0.535276,1.000000
75%,0.000000,1.000000,2.025000e+05,8.086500e+05,34596.000000,6.795000e+05,-12413.000000,-289.000000,-2010.000000,-1720.000000,3.000000,0.675053,6.636171e-01,0.669057,1.000000
max,1.000000,19.000000,1.170000e+08,4.050000e+06,258025.500000,4.050000e+06,-7489.000000,365243.000000,0.000000,0.000000,20.000000,0.962693,8.549997e-01,0.896010,1.000000


In [ ]:
# DESCRIBE DE LOS CAMPOS NO NUMERICOS

In [16]:
df.select_dtypes(include='object').describe()

,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,OCCUPATION_TYPE
count,307511,307511,307511,307511,307511,307511,307511,307511,211120
unique,2,3,2,2,8,5,6,6,18
top,Cash loans,F,N,Y,Working,Secondary / secondary special,Married,House / apartment,Laborers
freq,278232,202448,202924,213312,158774,218391,196432,272868,55186


Debemos analizar estas variables para buscar errores de tipeo y normalizarlas